auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/external/DeltaSetPasswordController.kt

```diff
@@ -0,0 +1,199 @@
+package uk.gov.communities.delta.auth.controllers.external
+
+import com.google.common.base.Strings
+import io.ktor.http.*
+import io.ktor.server.application.*
+import io.ktor.server.request.*
+import io.ktor.server.response.*
+import io.ktor.server.routing.*
+import io.ktor.server.thymeleaf.*
+import org.slf4j.LoggerFactory
+import uk.gov.communities.delta.auth.config.AuthServiceConfig
+import uk.gov.communities.delta.auth.config.DeltaConfig
+import uk.gov.communities.delta.auth.config.EmailConfig
+import uk.gov.communities.delta.auth.config.LDAPConfig
+import uk.gov.communities.delta.auth.plugins.UserVisibleServerError
+import uk.gov.communities.delta.auth.services.*
+import uk.gov.communities.delta.auth.utils.PasswordChecker
+
+class DeltaSetPasswordController(
+    private val deltaConfig: DeltaConfig,
+    private val ldapConfig: LDAPConfig,
+    private val emailConfig: EmailConfig,
+    private val authServiceConfig: AuthServiceConfig,
+    private val userService: UserService,
+    private val setPasswordTokenService: SetPasswordTokenService,
+    private val userLookupService: UserLookupService,
+    private val emailService: EmailService,
+) {
+    private val logger = LoggerFactory.getLogger(this.javaClass)
+    private val passwordChecker = PasswordChecker()
+
+    fun setPasswordFormRoutes(route: Route) {
+        route.post {
+            setPasswordPost(call)
+        }
+        route.get {
+            setPasswordGet(call)
+        }
+    }
+
+    fun setPasswordSuccessRoute(route: Route) {
+        route.get {
+            call.respondSuccessPage()
+        }
+    }
+
+    private suspend fun setPasswordGet(call: ApplicationCall) {
+        val userCN = call.request.queryParameters["userCN"].orEmpty()
+        val token = call.request.queryParameters["token"].orEmpty()
+        when (setPasswordTokenService.useToken(token, userCN)) {
+            is SetPasswordTokenService.NoSuchToken -> {
+                throw SetPasswordException("set_password_no_token", "Set password token did not exist")
+            }
+
+            is SetPasswordTokenService.ExpiredToken -> {
+                sendNewSetPasswordLink(userCN)
+                call.respondExpiredTokenPage()
```

Comment:
Sorry to make more work, but can you make sending a new activation email a button on the expired token page and have a separate endpoint for it? I think that will be clearer, and don't like side effects as part of GET requests